lib/vector-core/src/transforms/mod.rs

```diff
@@ -0,0 +1,125 @@
+    /// regard for coordination.
+    ///
+    /// **Note:** You should prefer to implement this over [`TaskTransform`]
+    /// where possible.
+    pub fn function(v: impl FunctionTransform + 'static) -> Self {
+        Transform::Function(Box::new(v))
+    }
+
+    /// Mutably borrow the inner transform as a function transform.
+    ///
+    /// # Panics
+    ///
+    /// If the transform is a [`TaskTransform`] this will panic.
+    pub fn as_function(&mut self) -> &mut Box<dyn FunctionTransform> {
+        match self {
+            Transform::Function(t) => t,
+            Transform::Task(_) => panic!(
+                "Called `Transform::as_function` on something that was not a function variant."
+            ),
+        }
+    }
+
+    /// Transmute the inner transform into a function transform.
+    ///
+    /// # Panics
+    ///
+    /// If the transform is a [`TaskTransform`] this will panic.
+    pub fn into_function(self) -> Box<dyn FunctionTransform> {
+        match self {
+            Transform::Function(t) => t,
+            Transform::Task(_) => panic!(
+                "Called `Transform::into_function` on something that was not a function variant."
+            ),
+        }
+    }
+
+    /// Create a new task transform.
+    ///
+    /// These tasks are coordinated, and map a stream of some `U` to some other
+    /// `T`.
+    ///
+    /// **Note:** You should prefer to implement [`FunctionTransform`] over this
+    /// where possible.
+    pub fn task(v: impl TaskTransform + 'static) -> Self {
+        Transform::Task(Box::new(v))
+    }
+
+    /// Mutably borrow the inner transform as a task transform.
+    ///
+    /// # Panics
+    ///
+    /// If the transform is a [`FunctionTransform`] this will panic.
+    pub fn as_task(&mut self) -> &mut Box<dyn TaskTransform> {
+        match self {
+            Transform::Function(_) => {
+                panic!("Called `Transform::as_task` on something that was not a task variant.")
+            }
+            Transform::Task(t) => t,
+        }
+    }
+
+    /// Transmute the inner transform into a task transform.
+    ///
+    /// # Panics
+    ///
+    /// If the transform is a [`FunctionTransform`] this will panic.
+    pub fn into_task(self) -> Box<dyn TaskTransform> {
+        match self {
+            Transform::Function(_) => {
+                panic!("Called `Transform::into_task` on something that was not a task variant.")
+            }
+            Transform::Task(t) => t,
+        }
+    }
+}
+
+/// Transforms that are simple, and don't require attention to coordination.
+/// You can run them as simple functions over events in any order.
+///
+/// # Invariants
+///
+/// * It is an illegal invariant to implement `FunctionTransform` for a
+///   `TaskTransform` or vice versa.
+pub trait FunctionTransform: Send + dyn_clone::DynClone + Sync {
+    fn transform(&mut self, output: &mut Vec<Event>, event: Event);
+}
```

Comment:
The `transform_one` that was previously here was marked test-only, which we can't export from the crate. I toyed with making a `FunctionTransformExt` but quickly felt that was overly complicated for what ends up being a helper function for tests.

Reply:
OTOH adding a `trait FunctionTransformTestExt` reduces all of the call site changes to an `import` addition. Whatever. It's all test code, so not a big deal.

Reply:
That's true. I couldn't figure out how to introduce that `Ext` without needing to sprinkle the code with impls, which felt noisier to me than adjusting how `transform_one` was called. I'd fully buy there's some fancy type work that could be done to avoid that.